stablehlo/reference/Ops.cpp

```diff
@@ -665,6 +796,58 @@ SmallVector<Tensor> evalReduceOp(ArrayRef<Tensor> inputs,
   return results;
 }
 
+SmallVector<Tensor> evalReduceWindowOp(
+    ArrayRef<Tensor> inputs, ArrayRef<Tensor> initValues,
+    const Sizes &windowDimensions, const Sizes &windowStrides,
+    const Sizes &baseDilations, const Sizes &windowDilations,
+    ArrayRef<std::pair<int64_t, int64_t>> padding, Region &body, Scope &scope,
+    ArrayRef<TensorType> resultTypes) {
+  SmallVector<Tensor> runtimeResults;
+
+  for (auto [resultType, initValue] : llvm::zip(resultTypes, initValues))
+    runtimeResults.push_back(Tensor(resultType, initValue.get({})));
+
+  SmallVector<int64_t> paddingLow;
+  for (auto paddingPair : padding) paddingLow.push_back(paddingPair.first);
+
+  SmallVector<Tensor> paddedInputs;
+  for (auto [input, initValue] : llvm::zip(inputs, initValues)) {
+    auto inferredPadType =
+        inferPadOpType(padding, input.getType(), initValue.getElementType(),
+                       SmallVector<int64_t>(baseDilations));
+    auto paddedInput = evalPadOp(input, initValue, Sizes(paddingLow),
+                                 baseDilations - 1, inferredPadType);
+    paddedInputs.push_back(paddedInput);
+  }
+
+  for (auto resultIt = runtimeResults[0].index_begin();
+       resultIt != runtimeResults[0].index_end(); ++resultIt) {
+    SmallVector<Tensor> windows;
+    auto windowStart = (*resultIt) * windowStrides;
+    for (auto paddedInput : paddedInputs) {
+      auto limitIndices = llvm::to_vector(windowStart + windowDimensions);
+      auto inferredSliceType =
+          inferSliceOpType(paddedInput.getType(), llvm::to_vector(windowStart),
+                           limitIndices, llvm::to_vector(windowDilations));
+      auto window = evalSliceOp(paddedInput, windowStart, windowDilations,
+                                inferredSliceType);
+      windows.push_back(window);
+    }
+
+    SmallVector<int64_t> dimensions(inputs[0].getRank());
+    std::iota(dimensions.begin(), dimensions.end(), 0);
+
+    SmallVector<TensorType> inferredReduceTypes =
+        inferReduceOpType(inputs, initValues, dimensions);
+    auto reducedValues = evalReduceOp(windows, initValues, Axes(dimensions),
+                                      body, scope, inferredReduceTypes);
+
+    for (auto [result, value] : llvm::zip(runtimeResults, reducedValues))
+      result.set(*resultIt, value.get({}));
+  }
+  return runtimeResults;
+}
```

Comment:
Thank you! It must've been a lot of work to figure this out, but it is satisfying that the implementation ended up being quite close to the spec. It is an open question whether this implementation will be sufficiently fast though, but time will tell.

Most of my feedback for this pull request involves simplifying the notation of `evalReduceWindowOp`, so that it more clearly maps on the spec. Here's what I ended up with (see https://github.com/ghpvnist/stablehlo/pull/3 for the implementation):

```C++
SmallVector<Tensor> evalReduceWindowOp(
    ArrayRef<Tensor> inputs, ArrayRef<Tensor> initValues,
    const Sizes &windowDimensions, const Sizes &windowStrides,
    const Sizes &baseDilations, const Sizes &windowDilations,
    const Sizes &paddingLow, const Sizes &paddingHigh, Region &body,
    Scope &scope, ArrayRef<TensorType> resultTypes) {
  SmallVector<Tensor> results;
  for (auto [resultType, initValue] : llvm::zip(resultTypes, initValues)) {
    results.push_back(Tensor(resultType, initValue.get({})));
  }

  SmallVector<Tensor> paddedInputs;
  for (auto [input, initValue] : llvm::zip(inputs, initValues)) {
    paddedInputs.push_back(evalPadOp(input, initValue, paddingLow, paddingHigh,
                                     baseDilations - 1));
  }

  for (auto resultIt = results[0].index_begin();
       resultIt != results[0].index_end(); ++resultIt) {
    SmallVector<Tensor> windows;
    auto windowStart = (*resultIt) * windowStrides;
    for (auto paddedInput : paddedInputs) {
      windows.push_back(evalSliceOp(paddedInput, windowStart,
                                    windowStart + windowDimensions,
                                    windowDilations));
    }

    Axes dimensions(inputs[0].getRank());
    std::iota(dimensions.begin(), dimensions.end(), 0);
    auto reducedValues =
        evalReduceOp(windows, initValues, dimensions, body, scope);
    for (auto [result, value] : llvm::zip(results, reducedValues))
      result.set(*resultIt, value.get({}));
  }
  return results;
}
```

Integrating inferFooOpType functions into evalFooOp overloads removes clutter and provides a cleaner look for evalReduceWindowOp, but that will have performance implications since each invocation of these overloads will perform type inference over and over. I would propose to move forward with a simpler implementation and then optimize later if it becomes a problem.